new/resource-actions.md

```diff
@@ -0,0 +1,147 @@
+end
+```
+
+There is no change to providers or the way we run actions.  This is just a way
+of defining a provider without having to type or think about the word.  Providers
+do not go away, so much as become a secondary concept, a sort of house for the
+primary concept of actions.
+
+`action` will work for any resource extending from `Chef::Resource`.
+
+### Action details
+
+#### Action recipe execution
+
+The action recipe works like a Chef run, in that it is an isolated resource
+collection and runs actions and notifications the same way as a top level Chef
+run.
+
+- The entire action recipe is compiled (just like a normal recipe) before
+  converging.
+- This compile happens when the action is *run*, not when the parent resource
+  is declared.
+- Delayed and immediate notifications are local to the action.
+
+This is equivalent to what `use_inline_resources` in an LWRP does, and we will
+keep them orthogonal as much as possible.
+
+#### Action composition and inheritance
+
+The action recipes are all added to a `Provider` class which inherits from the
+parent resource's provider class.  You may call `super` to call the parent
+class's version of the current action.
+
+```ruby
+class MyFile < Chef::Resource::File
+  action :create do
+    puts "I am a file yo"
+    super()
+  end
+end
+```
+
+### inline_recipe resource
+
+People need to support test-and-set and why-run in their resources. Given that
+actions are defined as recipes, the typical (and best) way to do that is
+resources (which have `not_if` and `only_if`, etc.). These resources work well,
+but sometimes you want to do a clump of actions based on a single `if`, or you
+want to nest them.
+
+`inline_recipe` is designed to do that job.  resource lets you write recipe
+code. It also has the parent in scope (so you can still access instance
+variables). The resources inside will compile immediately and be available in
+`inline_recipe.resources`.  The resources will converge when the `inline_recipe`
+converges.
+
+Like recipe actions, the resources inside the `inline_recipe` are not
+referenceable by `notifies` outside.  Unlike recipe actions, the resources show
+up in nested output (because the user typed them in the recipe).
+
+```ruby
+inline_recipe 'Create that one file' do
+  only_if { File.exist?('/please_create_that_file.txt') }
+
+  directory '/x' do
+  end
+  file '/x/y.txt' do
+  end
+end
+```
+
+`converge_by` (the current way of converging / updating) a resource will still
+be available for those who are used to it.
+
+## Changes to existing things
+
+### Resource
+
+We add the following public API to `Resource`:
+
+- `self.action(action, class: nil, &block)`: creates a Provider with the given
+  action
+
+And modify the following:
+
+- `allowed_actions`: defaults to all actions in the resource's provider class(es).
+- `default_action`: defaults to `allowed_actions.first`.
+
+### Provider
+
+Resources in LWRP actions will no longer report in output as nested actions
```

Comment:
wait, so if I make an LWRP that has 5 recources inside of it, those resources never get printed out? That'll make debugging my LWRP or understanding what went wrong VERY hard. :-1: